cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleServiceTest.java

```diff
@@ -452,6 +473,102 @@ public void testCreateAppDetailsArchive() throws Exception {
+
+    applicationLifecycleService.markAppsAsLatest(
+        Namespace.DEFAULT.toEntityId(),
+        new MarkLatestAppsRequest(
+            Collections.singletonList(new AppVersion(appName, ApplicationId.DEFAULT_VERSION))));
+
+    ApplicationDetail appDetail = applicationLifecycleService.getLatestAppDetail(appId.toEntityId().getAppReference());
+    Assert.assertNotNull(appDetail);
+    Assert.assertEquals(appName, appDetail.getName());
+    Assert.assertEquals(ApplicationId.DEFAULT_VERSION, appDetail.getAppVersion());
+
+    setLcmEditFlag(true);
+  }
+
+  @Test(expected = BadRequestException.class)
+  public void testMarkAppsAsLatestWithDuplicateAppIds() throws Exception {
+    applicationLifecycleService.markAppsAsLatest(
+        Namespace.DEFAULT.toEntityId(),
+        new MarkLatestAppsRequest(
+            Arrays.asList(new AppVersion("app1", "v1"),
+                new AppVersion("app1", "v2"))
+        ));
+  }
+
+  @Test(expected = BadRequestException.class)
+  public void testMarkAppsAsLatestWithInvalidAppId() throws Exception {
+    applicationLifecycleService.markAppsAsLatest(
+        Namespace.DEFAULT.toEntityId(),
+        new MarkLatestAppsRequest(
+            Collections.singletonList(new AppVersion("invalid app id", "v1"))
+        ));
+  }
+
+  @Test(expected = BadRequestException.class)
+  public void testMarkAppsAsLatestWithNullAppId() throws Exception {
+    applicationLifecycleService.markAppsAsLatest(
+        Namespace.DEFAULT.toEntityId(),
+        new MarkLatestAppsRequest(
+            Collections.singletonList(new AppVersion(null, "v1"))
+        ));
+  }
+
+  @Test(expected = BadRequestException.class)
+  public void testMarkAppsAsLatestWithNullAppVersion() throws Exception {
+    applicationLifecycleService.markAppsAsLatest(
+        Namespace.DEFAULT.toEntityId(),
+        new MarkLatestAppsRequest(
+            Collections.singletonList(new AppVersion("app", null))
+        ));
+  }
+
+  @Test(expected = ApplicationNotFoundException.class)
+  public void testMarkAppsAsLatestWithNonExistingAppVersion() throws Exception {
+    // disable the LCM edit flag for this test to use the default version for the app being deployed.
+    setLcmEditFlag(false);
+
+    String appName = "existing_app";
+    Id.Application appId = Id.Application.from(Id.Namespace.DEFAULT, appName);
+    Id.Artifact artifactId = Id.Artifact.from(Id.Namespace.DEFAULT, "appWithConfig", "1.0.0-SNAPSHOT");
+    addAppArtifact(artifactId, ConfigTestApp.class);
+    HttpResponse resp = deployWithoutMarkingLatest(
+        appId, new AppRequest<>(ArtifactSummary.from(artifactId.toArtifactId())));
+    Assert.assertEquals(200, resp.getResponseCode());
+
+    applicationLifecycleService.markAppsAsLatest(
+        Namespace.DEFAULT.toEntityId(),
+        new MarkLatestAppsRequest(
+            Arrays.asList(new AppVersion("non_existing_app", "v1"),
+                new AppVersion(appName, ApplicationId.DEFAULT_VERSION))
+        ));
+  }
+
```

Comment:
We should also have a partial failure test

Reply:
We do have a partial failure testcase. Now I have added javadoc comments on each of the test methods, so that it's easier to check the intended behaviour (verification) of each test.

Reply:
Also added the testcase that we discussed in the meeting (with 2 versions of the same application deployed one after another, then the later/second version marked as latest).